staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go

```diff
@@ -0,0 +1,287 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog/v2"
+)
+
+// TimingRatioHistogram is essentially a gauge for a ratio where the client
+// independently controls the numerator and denominator.
+// When scraped it produces a histogram of samples of the ratio
+// taken at the end of every nanosecond.
+// `*TimingRatioHistogram` implements both Registerable and RatioedGauge.
+type TimingRatioHistogram struct {
+	// The implementation is layered on TimingHistogram,
+	// adding the division by an occasionally adjusted denominator.
+
+	// Registerable is the registerable aspect.
+	// That is the registerable aspect of the underlying TimingHistogram.
+	compbasemetrics.Registerable
+
+	// timingRatioHistogramInner implements the RatioedObsserver aspect.
+	timingRatioHistogramInner
+}
+
+// TimingRatioHistogramOpts is the constructor parameters of a TimingRatioHistogram.
+// The `TimingHistogramOpts.InitialValue` is the initial numerator.
+type TimingRatioHistogramOpts struct {
+	compbasemetrics.TimingHistogramOpts
+	InitialDenominator float64
+}
+
+// timingRatioHistogramInner implements the instrumentation aspect
+type timingRatioHistogramInner struct {
+	nowFunc  func() time.Time
+	delegate interface{ Set(float64) }
```

Comment:
if we need an interface, can we define that instead of inlining it here?

Reply:
We could.  It would only be used internally to this one file, so I saw little value in defining it.

Reply:
Since you asked for it, I made it.